src/main/scala/components/memory-ports.scala

```diff
@@ -0,0 +1,229 @@
+ */
+class DMemPort(val blockwidth: Int) extends Module {
+  val io = IO (new DMemPortIO(blockwidth))
+  io := DontCare
+  io.request.valid  := false.B
+
+  val storedWrite = RegInit(0.U.asTypeOf(Valid(new PartialWrite(blockwidth))))
+  val memReallyReady = io.request.ready && !storedWrite.valid
+  io.ready := memReallyReady
+
+  // When the backing memory is ready and the pipeline is supplying a valid read OR write request, send out the request
+  // ... on the condition that there isn't a stored write in the queue.
+  // We need to process stored writes first to guarantee atomicity of the memory write operation
+
+  when (io.valid && memReallyReady && io.memread =/= io.memwrite) {
+    when (io.memwrite) {
+      storedWrite.bits.address   := io.address
+      storedWrite.bits.writedata := io.writedata
+      storedWrite.bits.maskmode  := io.maskmode
+      storedWrite.valid := true.B
+    }
+    
+    io.request.bits.address   := io.address
+    io.request.bits.writedata := 0.U(blockwidth.W)
+    io.request.bits.operation := Read
+    io.request.valid          := true.B
+  } .otherwise {
+    io.request.valid := false.B
+  }
+
+  // When memory is outputting data we need to determine whether it's to write back to memory or for simply
+  // reading
+  // This can be deduced from whether storedWrite is valid and the memory is signalling if it is ready.
+  when (io.response.valid) {
+    when (storedWrite.valid && io.request.ready) {
+      val writedata = Wire (UInt (blockwidth.W))
+
+      // When not writing a whole word
+      when (storedWrite.bits.maskmode =/= 2.U) {
+        // Read in the existing piece of data at the address, so we "overwrite" only part of it
+        val offset = storedWrite.bits.address (1, 0)
+        val readdata = Wire (UInt (blockwidth.W))
+        readdata := io.response.bits.data
+        val data = Wire (UInt (blockwidth.W))
+        // Mask the portion of the existing data so it can be or'd with the writedata
+        when (storedWrite.bits.maskmode === 0.U) {
+          data := io.response.bits.data & ~(0xff.U << (offset * 8.U))
+        } .otherwise {
+          data := io.response.bits.data & ~(0xffff.U << (offset * 8.U))
+        }
+        writedata := data | (storedWrite.bits.writedata << (offset * 8.U))
+      } .otherwise {
+        // Write the entire word
+        writedata := storedWrite.bits.writedata 
+      }
+
+      // Program the memory to issue a write
+      val request = Wire (new Request (blockwidth))
+      request.address   := storedWrite.bits.address
+      request.writedata := writedata
+      request.operation := Write
+      io.request.bits  := request
+      io.request.valid := true.B
+
+      // Mark the stored write register as being invalid.
+      storedWrite.valid := false.B
+    } .otherwise {
+      // Perform masking and sign extension on read data when memory is outputting it
+      val readdata_mask      = Wire(UInt(blockwidth.W))
```

Comment:
I added this assert but it failed whenever the data port was used. I also removed the `&& io.request.ready` condition to see if it was even needed, but this ended up failing the write tests.

So there is a case where `io.response.valid && ~io.request.ready` is true and corresponds to something meaningful in the mechanism, but I can't think of what it could be.

Reply:
Maybe I understand... If you have a stored write and you get a response from memory for a read, you should do this?

Or, should we just add an assert *here* that stored write isn't valid? 

Overall, this logic concerns me. I'd like to understand what we're expecting here.

Reply:
Nevermind, I found that it's a bug, and io.response.valid keeps being driven to be high since it was set to be DontCare initially. However fixing it causes all of the data port-related tests to fail, so I'll have to take a bit of time to figure out the deeper issue.

Reply:
Ah, I figured it out. It isn't a bug but rather just a fundamental feature of how the memory is designed.

The memory's `request.ready` output is hooked up to to the negation of `imemBusy` or `dmemBusy`. However both of these are registers, so their value doesn't change until the next clock cycle.

So when the memory executes a request, it sets `imemBusy` to false and `response.valid` to true. When both instruction memory and data memory reach the `when (io.response.valid)` line, they see (combinationally) that `io.response.valid` is true and `io.request.ready` is false, since `imemBusy` hasn't updated yet.

The quickest fix I can think of for this is to feed the read response of memory into a register at the same time `imemBusy` or `dmemBusy`, so that on the next clock cycle both `io.response.valid` and `io.request.ready` update at the same time.

Reply:
I don't quite follow this, but it sounds like you're starting to get a better grasp on it :).

Could you sketch something out on a piece of paper real quick which contains the design you have right now? I think it would be good to use to discuss this today in the meeting. Something kind of like the circuit diagrams we have for DINO CPU would be great which contains the registers and wires. No need to use any drawing software, pen+paper is perfect.